aiohttp/web_app.py

```diff
@@ -110,15 +110,18 @@ def freeze(self):
             return
 
         self._frozen = True
-        self._middlewares = tuple(self._prepare_middleware())
+        self._middlewares.freeze()
         self._router.freeze()
         self._on_response_prepare.freeze()
         self._on_startup.freeze()
         self._on_shutdown.freeze()
         self._on_cleanup.freeze()
+        self._run_middlewares = len(self.middlewares) > 0
+        self._middlewares_handlers = tuple(self._prepare_middleware())
 
         for subapp in self._subapps:
             subapp.freeze()
+            self._run_middlewares |= len(subapp._middlewares) > 0
```

Comment:
The line is cryptic a little.
Please replace with `self._run_middlewares = self._run_middlewares or subapp._run_middlewares`
The change respects middlewares from deep nested subapps: app -> subapp1 -> subapp2.

Reply:
Fair enough, done. Added also a bit of internal documentation about this flag and the root cause of its usage.